File: C04_InnerTube/InnerTubeVB/SharedUtilities/InnerTubeService.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.Xml
Imports System.Linq
Imports System.Net
Imports System.Diagnostics
Imports System.Xml.Linq
Imports System.Collections.Specialized
Imports System.Web

Namespace SharedUtilities
    Public Enum InnerTubeTime
        today
        this_week
        this_month
        all_time
    End Enum

    Public Class InnerTubeService
#Region "YouTube URL Properties"
        'Top Rated Videos
        Private Shared _TopRatedUrl As String = "http://gdata.youtube.com/feeds/api/standardfeeds/top_rated"
        Public Shared ReadOnly Property TopRatedUrl() As String
            Get
                Return _TopRatedUrl
            End Get
        End Property

        'Top Favorited Videos
        Private Shared _TopFavoritesUrl As String = "http://gdata.youtube.com/feeds/api/standardfeeds/top_favorites"
        Public Shared ReadOnly Property TopFavoritesUrl() As String
            Get
                Return _TopFavoritesUrl
            End Get
        End Property

        'Most Viewed
        Private Shared _MostViewedUrl As String = "http://gdata.youtube.com/feeds/api/standardfeeds/most_viewed"
        Public Shared ReadOnly Property MostViewedUrl() As String
            Get
                Return _MostViewedUrl
            End Get
        End Property

        'Favorites by User
        Private Shared _FavoritesByUserUrl As String = "http://gdata.youtube.com/feeds/api/users/username/favorites"
        Public Shared ReadOnly Property FavoritesByUserUrl() As String
            Get
                Return _FavoritesByUserUrl
            End Get
        End Property

        'Subscriptions By User
        Private Shared _SubscriptionsByUserUrl As String = "http://gdata.youtube.com/feeds/api/users/username/subscriptions"
        Public Shared ReadOnly Property SubscriptionsByUserUrl() As String
            Get
                Return _SubscriptionsByUserUrl
            End Get
        End Property

        'Search
        Private Shared _SearchUrl As String = "http://gdata.youtube.com/feeds/api/videos"
        Public Shared ReadOnly Property SearchUrl() As String
            Get
                Return _SearchUrl
            End Get
        End Property

        'baseUrl Embed Url, just append videoID
        Private Shared _baseEmbedUrl As String = "http://www.youtube.com/v/"
        Public Shared ReadOnly Property BaseEmbedUrl() As String
            Get
                Return _baseEmbedUrl
            End Get
        End Property

        'baseUrl Watch Url, just append videoID
        Private Shared _BasewatchUrl As String = "http://www.youtube.com/watch?v="
        Public Shared ReadOnly Property BaseWatchUrl() As String
            Get
                Return _BasewatchUrl
            End Get
        End Property

        'baseUrl Download Url, just append videoID
        Private Shared _BaseDownloadUrl As String = "http://www.youtube.com/get_video?video_id="
        Public Shared ReadOnly Property BaseDownloadUrl() As String
            Get
                Return _BaseDownloadUrl
            End Get
        End Property

        'baseUrl Thumbnail Url, just append videoID
        Private Shared _BaseThumbnailUrl As String = "http://img.youtube.com/vi/"
        Public Shared ReadOnly Property BaseThumbnailUrl() As String
            Get
                Return _BaseThumbnailUrl
            End Get
        End Property


        'Single Video Url, just append videoID
        Private Shared _SingleVideoUrl As String = "http://gdata.youtube.com/feeds/api/videos/"
        Public Shared ReadOnly Property SingleVideoUrl() As String
            Get
                Return _SingleVideoUrl
            End Get
        End Property
#End Region

#Region "Public Methods"
        Public Shared Function GetTopRated() As ObservableCollection(Of InnerTubeVideo)
            Return ConvertYouTubeXmlToObjects(_TopRatedUrl)
        End Function

        Public Shared Function GetTopRated(ByVal time As InnerTubeTime) As ObservableCollection(Of InnerTubeVideo)
            Return ConvertYouTubeXmlToObjects(BuildTimeUrl(_TopRatedUrl, time))
        End Function

        Public Shared Function GetMostViewed() As ObservableCollection(Of InnerTubeVideo)
            Return ConvertYouTubeXmlToObjects(MostViewedUrl)
        End Function

        Public Shared Function GetMostViewed(ByVal time As InnerTubeTime) As ObservableCollection(Of InnerTubeVideo)
            Return ConvertYouTubeXmlToObjects(BuildTimeUrl(MostViewedUrl, time))
        End Function

        Public Shared Function GetTopFavorites() As ObservableCollection(Of InnerTubeVideo)
            Return ConvertYouTubeXmlToObjects(TopFavoritesUrl)
        End Function

        Public Shared Function GetTopFavorites(ByVal time As InnerTubeTime) As ObservableCollection(Of InnerTubeVideo)
            Return ConvertYouTubeXmlToObjects(BuildTimeUrl(TopFavoritesUrl, time))
        End Function

        Public Shared Function GetFavoritesByUser(ByVal userId As String) As ObservableCollection(Of InnerTubeVideo)
            Return ConvertYouTubeXmlToObjects(BuildUserUrl(_FavoritesByUserUrl, userId))
        End Function

        Public Shared Function GetSubscriptionsByUser(ByVal userId As String) As ObservableCollection(Of InnerTubeVideo)
            Return ConvertYouTubeXmlToObjects(BuildUserUrl(_SubscriptionsByUserUrl, userId))
        End Function

        Public Shared Function Search(ByVal query As String) As ObservableCollection(Of InnerTubeVideo)

            'TODO: Add url escaping for query
            Return ConvertYouTubeXmlToObjects(BuildSearchUrl(query))
        End Function

        Public Shared Function GetSingleVideoById(ByVal Id As String) As ObservableCollection(Of InnerTubeVideo)
            Return ConvertYouTubeXmlToObjects(_SingleVideoUrl & Id)
        End Function

        Public Shared Function GetSingleVideo(ByVal watchUrl As String) As ObservableCollection(Of InnerTubeVideo)
            Dim queryParam As String = "v"

            Dim urlParts As NameValueCollection = HttpUtility.ParseQueryString(watchUrl)

            If String.IsNullOrEmpty(urlParts(queryParam)) Then
                Throw New Exception("Unable to parse URL, please use format: http://www.youtube.com/watch?v=oBmbZmrsz6U")
            End If

            Return GetSingleVideoById(urlParts(queryParam))
        End Function

        Public Shared Function ConvertYouTubeXmlToObjects(ByVal youTubeUrl As String) As ObservableCollection(Of InnerTubeVideo)
            Return ConvertYouTubeXmlToObjects(New Uri(youTubeUrl))
        End Function

        Public Shared Function ConvertYouTubeXmlToObjects(ByVal youTubeUrl As Uri) As ObservableCollection(Of InnerTubeVideo)

            Dim nsBase As XNamespace = "http://www.w3.org/2005/Atom"
            Dim nsGData As XNamespace = "http://schemas.google.com/g/2005"
            Dim nsYouTube As XNamespace = "http://gdata.youtube.com/schemas/2007"

            'call service
            Dim wc As New WebClient()

            'Get Data
            Dim xr As New XmlTextReader(wc.OpenRead(youTubeUrl))
            Dim rawData As XDocument = XDocument.Load(xr)

            Dim query = From entry In rawData.Descendants(nsBase + "entry") _
               Select New InnerTubeVideo With _
               { _
                .Author = entry.Element(nsBase + "author").Element(nsBase + "name").Value, _
                .Categories = ParseCategories(entry.Elements(nsBase + "category")), _
                .Id = ParseID(entry.Element(nsBase + "id").Value), _
                .Published = DateTime.Parse(entry.Element(nsBase + "published").Value), _
                .Updated = DateTime.Parse(entry.Element(nsBase + "updated").Value), _
                .Title = entry.Element(nsBase + "title").Value, _
                .Description = entry.Element(nsBase + "content").Value, _
                .ThumbnailLink = _BaseThumbnailUrl & ParseID(entry.Element(nsBase + "id").Value) & "/0.jpg", _
                .Link = _BasewatchUrl & ParseID(entry.Element(nsBase + "id").Value), _
                .EmbedLink = _baseEmbedUrl & ParseID(entry.Element(nsBase + "id").Value), _
                .DownloadLink = _BaseDownloadUrl & ParseID(entry.Element(nsBase + "id").Value), _
                .Views = Integer.Parse(entry.Element(nsYouTube + "statistics").Attribute("viewCount").Value), _
                .AvgRating = Single.Parse(entry.Element(nsGData + "rating").Attribute("average").Value), _
                .NumRaters = Integer.Parse(entry.Element(nsGData + "rating").Attribute("numRaters").Value) _
               }

            Return query.ToObservableCollection()

        End Function

        Public Shared Function ConvertYouTubeXmlToObjects(ByVal youTubeUrl As Uri, ByVal setting As Setting) As ObservableCollection(Of InnerTubeVideo)

            Dim nsBase As XNamespace = "http://www.w3.org/2005/Atom"
            Dim nsGData As XNamespace = "http://schemas.google.com/g/2005"
            Dim nsYouTube As XNamespace = "http://gdata.youtube.com/schemas/2007"

            'Use to call service
            Dim wc As New WebClient()

            'Get Data
            Dim xr As New XmlTextReader(wc.OpenRead(youTubeUrl))
            Dim rawData As XDocument = XDocument.Load(xr)


            Dim query = From entry In rawData.Descendants(nsBase + "entry") _
               Select New InnerTubeVideo With _
               { _
                .Author = entry.Element(nsBase + "author").Element(nsBase + "name").Value, _
                .Categories = ParseCategories(entry.Elements(nsBase + "category")), _
                .Id = ParseID(entry.Element(nsBase + "id").Value), _
                .Published = DateTime.Parse(entry.Element(nsBase + "published").Value), _
                .Updated = DateTime.Parse(entry.Element(nsBase + "updated").Value), _
                .Title = entry.Element(nsBase + "title").Value, _
                .Description = entry.Element(nsBase + "content").Value, _
                .ThumbnailLink = _BaseThumbnailUrl & ParseID(entry.Element(nsBase + "id").Value) & "/0.jpg", _
                .Link = _BasewatchUrl & ParseID(entry.Element(nsBase + "id").Value), _
                .EmbedLink = _baseEmbedUrl & ParseID(entry.Element(nsBase + "id").Value), _
                .DownloadLink = _BaseDownloadUrl & ParseID(entry.Element(nsBase + "id").Value), _
                .Views = Integer.Parse(entry.Element(nsYouTube + "statistics").Attribute("viewCount").Value), _
                .AvgRating = Single.Parse(entry.Element(nsGData + "rating").Attribute("average").Value), _
                .NumRaters = Integer.Parse(entry.Element(nsGData + "rating").Attribute("numRaters").Value), _
                .DownloadedImage = FileHelper.BuildFileName(setting.SubPath, ParseID(entry.Element(nsBase + "id").Value), FileType.Image), _
                .DownloadedFlv = FileHelper.BuildFileName(setting.SubPath, entry.Element(nsBase + "title").Value, FileType.Flv), _
                .DownloadedMp4 = FileHelper.BuildFileName(setting.VideoPath, entry.Element(nsBase + "title").Value, FileType.Mp4), .DownloadedWmv = FileHelper.BuildFileName(setting.VideoPath, entry.Element(nsBase + "title").Value, FileType.Wmv) _
               }

            Return query.ToObservableCollection()

        End Function


#End Region

#Region "Private Methods"

        Private Shared Function ParseComments(ByVal p As String) As Integer 'not called
            If String.IsNullOrEmpty(p) Then
                Return -1
            Else
                Return Integer.Parse(p)
            End If
        End Function

        Public Shared Function BuildSearchUrl(ByVal query As String) As String
            Return _SearchUrl & "vq=" & query
        End Function

        Public Shared Function BuildTimeUrl(ByVal baseUrl As String, ByVal time As InnerTubeTime) As String
            Dim build As New UriBuilder(baseUrl)
            build.Query = "time=" & time.ToString()
            Return build.Uri.ToString()
        End Function

        Public Shared Function BuildUserUrl(ByVal url As String, ByVal userId As String) As String
            Dim val As String = url.Replace("username", userId)
            Return val
        End Function

        Private Shared Function ParseTime(ByVal seconds As String) As TimeSpan 'not called
            Dim timevalue As Integer = Integer.Parse(seconds)
            Debug.WriteLine("timevalue=" & timevalue)

            Try
                Dim t As New TimeSpan(0, 0, timevalue)
                Return t
            Catch ex As Exception
                Debug.WriteLine("msg" & ex.Message)
                Debug.WriteLine(ex.Source)
                Throw
            End Try

        End Function

        Private Shared Function ParseCategories(ByVal Categories As IEnumerable(Of XElement)) As Collection(Of String)
            Dim vals = From c In Categories.Attributes("term") _
                       Select c.Value
            Return vals.ToCollection()

        End Function

        Private Shared Function ParseID(ByVal url As String) As String
            'Format: "http://gdata.youtube.com/feeds/api/videos/cI1AwZN4ZYg";
            Dim x = url.Split("/"c)
            Return x.Last()
        End Function

#End Region

    End Class

End Namespace

File: C4_InnerTube/InnerTubeCS/SharedUtilities/InnerTubeService.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Xml;
using System.Linq;
using System.Net;
using System.Diagnostics;
using System.Xml.Linq;
using System.Web;
using System.Collections.Specialized;

namespace SharedUtilities
{
    public enum InnerTubeTime
    {
        today,
        this_week,
        this_month,
        all_time
    }

    public class InnerTubeService
    {
        #region YouTube URL Properties
        //Top Rated Videos
        private static string _TopRatedUrl = "http://gdata.youtube.com/feeds/api/standardfeeds/top_rated";
        public static string TopRatedUrl
        { get { return _TopRatedUrl; } }

        //Top Favorited Videos
        private static string _TopFavoritesUrl = "http://gdata.youtube.com/feeds/api/standardfeeds/top_favorites";
        public static string TopFavoritesUrl
        { get { return _TopFavoritesUrl; } }

        //Most Viewed
        private static string _MostViewedUrl = "http://gdata.youtube.com/feeds/api/standardfeeds/most_viewed";
        public static string MostViewedUrl
        { get { return _MostViewedUrl; } }

        //Favorites by User
        private static string _FavoritesByUserUrl = "http://gdata.youtube.com/feeds/api/users/username/favorites";
        public static string FavoritesByUserUrl
        { get { return _FavoritesByUserUrl; } }

        //Subscriptions By User
        private static string _SubscriptionsByUserUrl = "http://gdata.youtube.com/feeds/api/users/username/subscriptions";
        public static string SubscriptionsByUserUrl
        { get { return _SubscriptionsByUserUrl; } }

        //Search
        private static string _SearchUrl = "http://gdata.youtube.com/feeds/api/videos";
        public static string SearchUrl
        { get { return _SearchUrl; } }

        //baseUrl Embed Url, just append videoID
        private static string _baseEmbedUrl = "http://www.youtube.com/v/";
        public static string BaseEmbedUrl
        { get { return _baseEmbedUrl; } }

        //baseUrl Watch Url, just append videoID
        private static string _BasewatchUrl = "http://www.youtube.com/watch?v=";
        public static string BaseWatchUrl
        { get { return _BasewatchUrl; } }

        //baseUrl Download Url, just append videoID
        private static string _BaseDownloadUrl = "http://www.youtube.com/get_video?video_id=";
        public static string BaseDownloadUrl
        { get { return _BaseDownloadUrl; } }

        //baseUrl Thumbnail Url, just append videoID
        private static string _BaseThumbnailUrl = "http://img.youtube.com/vi/";
        public static string BaseThumbnailUrl
        { get { return _BaseThumbnailUrl; } }


        //Single Video Url, just append videoID
        private static string _SingleVideoUrl = "http://gdata.youtube.com/feeds/api/videos/";
        public static string SingleVideoUrl
        { get { return _SingleVideoUrl; } }
        #endregion

        #region Public Methods
        public static ObservableCollection<InnerTubeVideo> GetTopRated()
        {
            return ConvertYouTubeXmlToObjects(_TopRatedUrl);
        }

        public static ObservableCollection<InnerTubeVideo> GetTopRated(InnerTubeTime time)
        {
            return ConvertYouTubeXmlToObjects(BuildTimeUrl(_TopRatedUrl, time));
        }

        public static ObservableCollection<InnerTubeVideo> GetMostViewed()
        {
            return ConvertYouTubeXmlToObjects(MostViewedUrl);
        }

        public static ObservableCollection<InnerTubeVideo> GetMostViewed(InnerTubeTime time)
        {
            return ConvertYouTubeXmlToObjects(BuildTimeUrl(MostViewedUrl, time));
        }

        public static ObservableCollection<InnerTubeVideo> GetTopFavorites()
        {
            return ConvertYouTubeXmlToObjects(TopFavoritesUrl);
        }

        public static ObservableCollection<InnerTubeVideo> GetTopFavorites(InnerTubeTime time)
        {
            return ConvertYouTubeXmlToObjects(BuildTimeUrl(TopFavoritesUrl, time));
        }

        public static ObservableCollection<InnerTubeVideo> GetFavoritesByUser(string userId)
        {
            return ConvertYouTubeXmlToObjects(BuildUserUrl(_FavoritesByUserUrl, userId));
        }

        public static ObservableCollection<InnerTubeVideo> GetSubscriptionsByUser(string userId)
        {
            return ConvertYouTubeXmlToObjects(BuildUserUrl(_SubscriptionsByUserUrl, userId));
        }

        public static ObservableCollection<InnerTubeVideo> Search(string query)
        {

            //TODO: Add url escaping for query
            return ConvertYouTubeXmlToObjects(BuildSearchUrl(query));
        }

        public static ObservableCollection<InnerTubeVideo> GetSingleVideoById(string Id)
        {
            return ConvertYouTubeXmlToObjects(_SingleVideoUrl + Id);
        }

        public static ObservableCollection<InnerTubeVideo> GetSingleVideo(string watchUrl)
        {
            string queryParam = "v";

            NameValueCollection urlParts = HttpUtility.ParseQueryString(watchUrl);

            if (string.IsNullOrEmpty(urlParts[queryParam]) )
            {
                throw new Exception(@"Unable to parse URL, please use format: http://www.youtube.com/watch?v=oBmbZmrsz6U");
            }

            return GetSingleVideoById(urlParts[queryParam]); 
        }

        public static ObservableCollection<InnerTubeVideo> ConvertYouTubeXmlToObjects(string youTubeUrl)
        {
            return ConvertYouTubeXmlToObjects(new Uri(youTubeUrl));
        }

        public static ObservableCollection<InnerTubeVideo> ConvertYouTubeXmlToObjects(Uri youTubeUrl)
        {

            XNamespace nsBase = @"http://www.w3.org/2005/Atom";
            XNamespace nsGData = @"http://schemas.google.com/g/2005";
            XNamespace nsYouTube = @"http://gdata.youtube.com/schemas/2007";

            //call service
            WebClient wc = new WebClient();

            //Get Data
            XmlTextReader xr = new XmlTextReader(wc.OpenRead(youTubeUrl));
            XDocument rawData = XDocument.Load(xr);

            var query = from entry in rawData.Descendants(nsBase + "entry")
                        select new InnerTubeVideo
                        {
                            Author = entry.Element(nsBase + "author").Element(nsBase + "name").Value,
                            Categories = ParseCategories(entry.Elements(nsBase + "category")),
                            Id = ParseID(entry.Element(nsBase + "id").Value),
                            Published = DateTime.Parse(entry.Element(nsBase + "published").Value),
                            Updated = DateTime.Parse(entry.Element(nsBase + "updated").Value),
                            Title = entry.Element(nsBase + "title").Value,
                            Description = entry.Element(nsBase + "content").Value,
                            ThumbnailLink = _BaseThumbnailUrl + ParseID(entry.Element(nsBase + "id").Value) + @"/0.jpg",
                            Link = _BasewatchUrl + ParseID(entry.Element(nsBase + "id").Value),
                            EmbedLink = _baseEmbedUrl + ParseID(entry.Element(nsBase + "id").Value),
                            DownloadLink = _BaseDownloadUrl + ParseID(entry.Element(nsBase + "id").Value),
                            Views = int.Parse(entry.Element(nsYouTube + "statistics").Attribute("viewCount").Value),
                            AvgRating = float.Parse(entry.Element(nsGData + "rating").Attribute("average").Value),
                            NumRaters = int.Parse(entry.Element(nsGData + "rating").Attribute("numRaters").Value),
                        };

            return query.ToObservableCollection<InnerTubeVideo>();

        }

        public static ObservableCollection<InnerTubeVideo> ConvertYouTubeXmlToObjects(Uri youTubeUrl, Setting setting)
        {

            XNamespace nsBase = @"http://www.w3.org/2005/Atom";
            XNamespace nsGData = @"http://schemas.google.com/g/2005";
            XNamespace nsYouTube = @"http://gdata.youtube.com/schemas/2007";

            //Use to call service
            WebClient wc = new WebClient();

            //Get Data
            XmlTextReader xr = new XmlTextReader(wc.OpenRead(youTubeUrl));
            XDocument rawData = XDocument.Load(xr);

            var query = from entry in rawData.Descendants(nsBase + "entry")
                        select new InnerTubeVideo
                        {
                            Author = entry.Element(nsBase + "author").Element(nsBase + "name").Value,
                            Categories = ParseCategories(entry.Elements(nsBase + "category")),
                            Id = ParseID(entry.Element(nsBase + "id").Value),
                            Published = DateTime.Parse(entry.Element(nsBase + "published").Value),
                            Updated = DateTime.Parse(entry.Element(nsBase + "updated").Value),
                            Title = entry.Element(nsBase + "title").Value,
                            Description = entry.Element(nsBase + "content").Value,
                            ThumbnailLink = _BaseThumbnailUrl + ParseID(entry.Element(nsBase + "id").Value) + @"/0.jpg",
                            Link = _BasewatchUrl + ParseID(entry.Element(nsBase + "id").Value),
                            EmbedLink = _baseEmbedUrl + ParseID(entry.Element(nsBase + "id").Value),
                            DownloadLink = _BaseDownloadUrl + ParseID(entry.Element(nsBase + "id").Value),
                            Views = int.Parse(entry.Element(nsYouTube + "statistics").Attribute("viewCount").Value),
                            AvgRating = float.Parse(entry.Element(nsGData + "rating").Attribute("average").Value),
                            NumRaters = int.Parse(entry.Element(nsGData + "rating").Attribute("numRaters").Value),
                            //set download locations
                            DownloadedImage = FileHelper.BuildFileName(setting.SubPath,
                                                ParseID(entry.Element(nsBase + "id").Value), FileType.Image),
                            DownloadedFlv = FileHelper.BuildFileName(setting.SubPath,
                                                entry.Element(nsBase + "title").Value, FileType.Flv),
                            DownloadedMp4 = FileHelper.BuildFileName(setting.VideoPath,
                                                entry.Element(nsBase + "title").Value, FileType.Mp4),
                            DownloadedWmv = FileHelper.BuildFileName(setting.VideoPath,
                                                entry.Element(nsBase + "title").Value, FileType.Wmv)
                        };

            return query.ToObservableCollection<InnerTubeVideo>();

        }


        #endregion

        #region Private Methods

        private static int ParseComments(string p)  //not called
        {
            if (String.IsNullOrEmpty(p))
            {
                return -1;
            }
            else
            {
                return int.Parse(p);
            }
        }

        public static string BuildSearchUrl(string query)
        {
            return _SearchUrl + "vq=" + query;
        }

        public static string BuildTimeUrl(string baseUrl, InnerTubeTime time)
        {
            UriBuilder build = new UriBuilder(baseUrl);
            build.Query = "time=" + time.ToString();
            return build.Uri.ToString();
        }

        public static string BuildUserUrl(string url, string userId)
        {
            string val = url.Replace("username", userId);
            return val;
        }

        private static TimeSpan ParseTime(string seconds) //not called
        {
            int timevalue = int.Parse(seconds);
            Debug.WriteLine("timevalue=" + timevalue);

            try
            {
                TimeSpan t = new TimeSpan(0, 0, timevalue);
                return t;
            }
            catch (Exception ex)
            {
                Debug.WriteLine("msg" + ex.Message);
                Debug.WriteLine(ex.Source);
                throw;
            }

        }

        private static Collection<string> ParseCategories(IEnumerable<XElement> Categories)
        {
            var vals = from c in Categories.Attributes("term")
                       select c.Value;
            return vals.ToCollection<string>();
        }

        private static string ParseID(string url)
        {
            //Format: "http://gdata.youtube.com/feeds/api/videos/cI1AwZN4ZYg";
            var x = url.Split('/');
            return x.Last();
        }

        #endregion

    }

}
